va/va.go

```diff
@@ -534,6 +556,80 @@ func (va VAImpl) fetchHTTP(identifier string, token string) ([]byte, string, *ac
 	return body, url.String(), nil
 }
 
+func (va VAImpl) getTXTEntry(name string) ([]string, error) {
+	ctx, cancelfunc := context.WithTimeout(context.Background(), validationTimeout)
+	defer cancelfunc()
+
+	if va.customResolverAddr == "" {
+		return net.DefaultResolver.LookupTXT(ctx, name)
+	}
+
+	txts := []string{}
+	client := new(dns.Client)
+	message := new(dns.Msg)
+	message.SetQuestion(dns.Fqdn(name), dns.TypeTXT)
+	in, _, err := client.ExchangeContext(ctx, message, va.customResolverAddr)
+
+	if err != nil || in.Rcode != dns.RcodeSuccess {
+		return txts, err
+	}
+
+	for _, record := range in.Answer {
+		if t, ok := record.(*dns.TXT); ok {
+			txts = append(txts, t.Txt...)
+		}
+	}
+
+	return txts, nil
+}
+
+func (va VAImpl) resolveIP(name string) ([]string, error) {
+	ctx, cancelfunc := context.WithTimeout(context.Background(), validationTimeout)
+	defer cancelfunc()
+
+	if va.customResolverAddr == "" {
+		return net.DefaultResolver.LookupHost(ctx, name)
+	}
+
+	addrs := []string{}
+	parsed := net.ParseIP(name)
+
+	if parsed != nil {
+		addrs = append(addrs, name)
+		return addrs, nil
+	}
+
+	client := new(dns.Client)
+	messageA := new(dns.Msg)
+	messageA.SetQuestion(dns.Fqdn(name), dns.TypeA)
+	inA, _, errA := client.ExchangeContext(ctx, messageA, va.customResolverAddr)
+	messageAAAA := new(dns.Msg)
+	messageAAAA.SetQuestion(dns.Fqdn(name), dns.TypeAAAA)
+	inAAAA, _, errAAAA := client.ExchangeContext(ctx, messageAAAA, va.customResolverAddr)
+
+	if errA != nil {
+		return addrs, errA
+	}
+
+	if errAAAA != nil {
+		return addrs, errAAAA
+	}
+
+	for _, record := range inA.Answer {
+		if t, ok := record.(*dns.A); ok {
+			addrs = append(addrs, t.A.String())
+		}
+	}
+
+	for _, record := range inAAAA.Answer {
```

Comment:
Are you open to changing the order of these two `ranges`? I think the AAAA records should appear first in the `addrs` slice to better match typical "happy eyeballs" behaviour.

The godoc comment for this function should also mention the order of results w.r.t. dual-homed names.

Reply:
Indeed, I think it would be beneficial for the sake of the future of IPv6. However to work properly, maybe it is needed to make some conditional appends here. I mean that on a non-IPv6 capable system, this function should not return any IPv6 addresses (and similarly for IPv4) right ?

If so, @cpu do you have an idea of how I could test the availability of each network stack on the current system, to conditionally trigger one, or both, of the two loops ?

Reply:
I think I would advocate an approach where we ignore the availability of the network stack. If a Pebble environment is configured with a DNS server that returns AAAA records for a name and the Pebble environment doesn't have IPv6 connectivity I think the solution is to remove the AAAA records from the mock data.

Reply:
That works for me. I switch the order of `AAAA` and `A` records here.

Reply:
Great, thanks! I'll give this branch another pass shortly.